File: 0146-lru-cache/0146-lru-cache.swift
//ChatGPT, thank you

class Node {
    var key: Int
    var value: Int
    var prev: Node?
    var next: Node?

    init(_ key: Int = 0, _ value: Int = 0) {
        self.key = key
        self.value = value
    }
}

class LRUCache {
    private let capacity: Int
    private var cache = [Int: Node]()
    private let head = Node()
    private let tail = Node()

    init(_ capacity: Int) {
        self.capacity = capacity
        head.next = tail
        tail.prev = head
    }

    private func remove(_ node: Node) {
        let prev = node.prev
        let next = node.next
        prev?.next = next
        next?.prev = prev
    }

    private func insertAtTail(_ node: Node) {
        let prev = tail.prev
        prev?.next = node
        node.prev = prev
        node.next = tail
        tail.prev = node
    }

    func get(_ key: Int) -> Int {
        guard let node = cache[key] else { return -1 }
        remove(node)
        insertAtTail(node)
        return node.value
    }

    func put(_ key: Int, _ value: Int) {
        if let node = cache[key] {
            node.value = value
            remove(node)
            insertAtTail(node)
        } else {
            if cache.count >= capacity, let lru = head.next, lru !== tail {
                remove(lru)
                cache.removeValue(forKey: lru.key)
            }

            let newNode = Node(key, value)
            cache[key] = newNode
            insertAtTail(newNode)
        }
    }
}

File: 0146-lru-cache/0146-lru-cache.py
class DoubleLinkedList:
    def __init__(self, key=0, value=0, nxt=None, prev=None):
        self.key = key
        self.value = value
        self.nxt = nxt
        self.prev = prev

class LRUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.cache = {}
        self.head = DoubleLinkedList()
        self.tail = DoubleLinkedList()
        self.head.nxt = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self.remove(node)
        self.insert(node)
        return node.value
    
    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            node = DoubleLinkedList(key, value)
            self.insert(node)
            self.cache[key] = node
        else:
            self.cache[key].value = value
            node = self.cache[key]
            self.remove(node)
            self.insert(node)

        if len(self.cache) > self.capacity:
            #Remove LRU
            del self.cache[self.head.nxt.key]
            self.remove(self.head.nxt)

    def remove(self, node):
        node.nxt.prev = node.prev
        node.prev.nxt = node.nxt

    def insert(self, node):
        temp = self.tail.prev
        self.tail.prev = node
        node.nxt = self.tail
        
        node.prev = temp
        temp.nxt = node
